cpp/src/search/contains_table.cu

```diff
@@ -1,148 +1,378 @@
-      using nan_equal_comparator =
-        cudf::experimental::row::equality::nan_equal_physical_equality_comparator;
-      check_contains(nan_equal_comparator{});
-    } else {
-      using nan_unequal_comparator =
-        cudf::experimental::row::equality::physical_equality_comparator;
-      check_contains(nan_unequal_comparator{});
-    }
-  }
-
-  return contained;
-}
-
-}  // namespace cudf::detail
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <join/join_common_utils.cuh>
+
+#include <cudf/detail/join.hpp>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/detail/structs/utilities.hpp>
+#include <cudf/table/experimental/row_operators.cuh>
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/table/table_view.hpp>
+#include <cudf/types.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
+
+#include <thrust/iterator/counting_iterator.h>
+
+#include <cuco/static_map.cuh>
+
+namespace cudf::detail {
+
+namespace {
+
+using cudf::experimental::row::lhs_index_type;
+using cudf::experimental::row::rhs_index_type;
+
+using static_map = cuco::static_map<lhs_index_type,
+                                    size_type,
+                                    cuda::thread_scope_device,
+                                    rmm::mr::stream_allocator_adaptor<default_allocator<char>>>;
+
+/**
+ * @brief An adapter functor to support strong index types for row hasher that must be operating on
+ * `cudf::size_type`.
+ */
+template <typename Hasher>
+struct strong_index_hasher_adapter {
+  strong_index_hasher_adapter(Hasher const& hasher) : _hasher{hasher} {}
+
+  template <typename T>
+  __device__ inline auto operator()(T const idx) const noexcept
+  {
+    return _hasher(static_cast<size_type>(idx));
+  }
+
+ private:
+  Hasher _hasher;
+};
+
+/**
+ * @brief An adapter functor to support strong index type for table row comparator that must be
+ * operating on `cudf::size_type`.
+ */
+template <typename Comparator>
+struct strong_index_comparator_adapter {
+  strong_index_comparator_adapter(Comparator const& comparator) : _comparator{comparator} {}
+
+  template <typename T>
+  __device__ inline auto operator()(T const lhs_index, T const rhs_index) const noexcept
+  {
+    return _comparator(static_cast<size_type>(lhs_index), static_cast<size_type>(rhs_index));
+  }
+
+  __device__ inline auto operator()(lhs_index_type const lhs_index,

```

Comment:
Can all these be `constexpr` instead of `inline`? [The `strong_index_comparator_adapter` uses `constexpr`.](https://github.com/rapidsai/cudf/blob/ebcea0fa1edba776a1631aac4b7e50b8127a3e4f/cpp/include/cudf/table/experimental/row_operators.cuh#L684)